src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas3/DomainEventService.kt

```diff
@@ -0,0 +1,135 @@
+
+import com.amazonaws.services.sns.model.MessageAttributeValue
+import com.amazonaws.services.sns.model.PublishRequest
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.events.cas3.model.CAS3Event
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.events.cas3.model.CAS3PersonArrivedEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventType
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.DomainEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventAdditionalInformation
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReference
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReferenceCollection
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
+import uk.gov.justice.hmpps.sqs.MissingTopicException
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.util.UUID
+import javax.transaction.Transactional
+import kotlin.reflect.KClass
+
+@Service(
+  "uk.gov.justice.digital.hmpps.approvedpremisesapi.service.cas3.DomainEventService",
+)
+class DomainEventService(
+  private val objectMapper: ObjectMapper,
+  private val domainEventRepository: DomainEventRepository,
+  private val hmppsQueueService: HmppsQueueService,
+  @Value("\${domain-events.cas3.emit-enabled}") private val emitDomainEventsEnabled: Boolean,
+  @Value("\${url-templates.api.cas3.person-arrived-event-detail}") private val personArrivedDetailUrlTemplate: String,
+) {
+  private val log = LoggerFactory.getLogger(this::class.java)
+
+  private val domainTopic by lazy {
+    hmppsQueueService.findByTopicId("domainevents")
+      ?: throw MissingTopicException("domainevents not found")
+  }
+
+  fun getPersonArrivedEvent(id: UUID) = get<CAS3PersonArrivedEvent>(id)
+
+  private inline fun <reified T : CAS3Event> get(id: UUID): DomainEvent<T>? {
+    val domainEventEntity = domainEventRepository.findByIdOrNull(id) ?: return null
+
+    val data = when {
+      enumTypeFromDataType(T::class) == domainEventEntity.type ->
+        objectMapper.readValue(domainEventEntity.data, T::class.java)
+      else -> throw RuntimeException("Unsupported DomainEventData type ${T::class.qualifiedName}/${domainEventEntity.type.name}")
+    }
+    return DomainEvent(
+      id = domainEventEntity.id,
+      applicationId = domainEventEntity.applicationId,
```

Comment:
Is applicationId our application/referral uuid? Just checking that we're still prepared for this to be null.

Should we send the bookingId here which is something we will always have?

Reply:
The application ID is required by `SnsEventAdditionalInformation` when the event is published, and as this is part of the data that's directly exposed to an external service that I know very little about I've opted not to make this nullable. If we don't have an application ID available we could either choose not to emit an event or to use a zero UUID as a stand-in, which is the approach I've taken here.

Reply:
Sounds like we need to figure this one out. I'm not sure what a zero UUID means but sending a stand in to a receiver who is expecting something else could cause problems.

[It seems like CAS1 added the domain event record and included applicationId on the entity](https://github.com/ministryofjustice/hmpps-approved-premises-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/DomainEvent.kt#L8) and they then presumably expose it in every(?) domain event, and have as a side affect made that tightly coupled?

CAS3 will be sending several domain events based on a booking and one based on an application(referral.submitted).

Assuming someone is consuming this data (and we should check with the integrations team) is what we need here a change to the domain events to make this polymorphic? 

```kotlin
data class DomainEvent<T> (
  val id: UUID,
  val recordId: UUID,
  val recordType: Booking | Application | Any?,
  val crn: String,
  val occurredAt: Instant,
  val data: T,
)
```

This would allow us to always send an identifier for the primary record of interest with a booking id or an application id together with the detail of type of record it belongs to.

Alternatively we make it nullable though we should also check if the receiver is ok with that as presumably they had a need for it originally.

Reply:
I'd expect the worst case scenario for using a zero UUID (`00000000-0000-0000-0000-000000000000`) would be that the service at the other end of the message queue would use it to do a request to the `GET /applications/{applicationId}` endpoint, which would then give them a 404 response.

This suggestion to overhaul what a domain event looks like is a larger breaking change of the contract between the API and the consumer of the queue. While it's probably a more accurate model, I wouldn't feel comfortable doing any further work on domain events until it's been agreed with all relevant teams, which would put us in significant danger of missing our deadline on this feature.

Reply:
[For reference the conversation is continuing in the Slack channel with the probation integrations team.](https://mojdt.slack.com/archives/C03KVQUFSER/p1696846088073159)